File: dataStructureStudy/LinkedList.py
class node :
	def __init__(self, data = None, pre = None, next = None):
		self.data = data;
		self.pre = pre;
		self.next = next;

class linkedList :
	def __init__(self) :
		self.start = node();
		self.end = node();
		self.start.pre = self.start;
		self.start.next = self.end;
		self.end.pre = self.start;
		self.size = 0;

	
	def getNindex(self, n) :
		if n >= self.size : 
			return self.end.pre;
		else : 
			currentnode = self.start;

		for i in range(n+1) : 
			currentnode = currentnode.next;
		
		return currentnode;
		
	def insert(self, value, n = -1) : 
		if self.size == 0 : 
			newNode = node(value, self.start, self.end);
			self.start.next = newNode;
			self.end.pre = newNode;
		else : 
			if n == -1 : #insert last place
				newNode = node(value, self.end.pre, self.end); 
				self.end.pre.next = newNode;
				self.end.pre = newNode;

			else :
				nextNode = self.getNindex(n);
				if n >= self.size : 
				    nextNode = self.end;
				preNode = nextNode.pre;
				newNode = node(value, preNode, nextNode);
				preNode.next = newNode;
				nextNode.pre = newNode;
	
		self.size = self.size + 1;
	def delete(self, n = -1) : 
		if n == -1 : 
			delNode = self.end.pre;
			self.end.pre.pre.next = self.end;
			self.end.pre = self.end.pre.pre
		else : 
			delNode = self.getNindex(n);
			delNode.pre.next = delNode.next;
			delNode.next.pre = delNode.pre;

	def printall(self, direction = True) :
		if direction : 
			currentnode = self.start.next;
			
			while currentnode != self.end : 
				print currentnode.data, 
				currentnode = currentnode.next;
		else :
			currentnode = self.end.pre;
			
			while currentnode != self.start : 
				print currentnode.data, 
				currentnode = currentnode.pre;
		
		print ""
	def swap(self, a, b) : 
		nodeA = self.getNindex(a);
		nodeB = self.getNindex(b);

		temp = nodeA.data;
		nodeA.data = nodeB.data;
		nodeB.data = temp;


ll = linkedList();
ll.insert(1);
ll.insert(2);
ll.insert(3);
ll.insert(4);
ll.insert(5);
ll.printall(True)
ll.printall(False)

print ""
ll.swap(0,66666)
ll.printall(True)
ll.printall(False)

File: dataStructureStudy/linkedList.cpp
#include <iostream>

using namespace std;
class node{
private:
public:
  int data;
  node* next;
  node* pre;
  node(){
    next = NULL;
    pre = NULL;
  }
  node(int data){
    this->data = data;
    next = NULL;
    pre = NULL;
  }

};

class linkedList{
private:
  int size;
  node* start;
  node* end;
public:
  linkedList(){
    size = 0;
    start = new node();
    end = new node();
    start->pre = start;

    end->next = end;
  }
  node* getNnode(int n){
    /*if(size < n){
      return end->pre;
    }*/
    node* index = start;
    for(int i = 0; i < n && index!=end->pre; i++){
      index = index->next;
    }

    return index;
  }

  linkedList* insert(int data){
    if(size == 0){
      node* insertNode = new node(data);
      start->next = insertNode;
      insertNode->pre = start;
      insertNode->next = end;
      end->pre = insertNode;
    }
    else{
      node* insertNode = new node(data);
      node* index = getNnode(size);
      node* indexNext = index->next;

      index->next = insertNode;
      indexNext->pre = insertNode;
      insertNode->next = indexNext;
      insertNode->pre = index;

    }
    size++;

    return this;
  }

  linkedList* insert(int data, int n){
    node* insertNode = new node(data);
    node* index = getNnode(n);
    node* indexNext = index->next;

    index->next = insertNode;
    indexNext->pre = insertNode;
    insertNode->next = indexNext;
    insertNode->pre = index;

    size++;

    return this;
  }

  linkedList* printAll(){
    node* index = start->next;
    for(; index!=end; index = index->next){
      cout << index->data <<" ";
    }
    cout << endl;

    index = end->pre;
    for(; index!=start; index = index->pre){
      cout << index->data <<" ";
    }
    cout << endl;
    return this;
  }



  linkedList* delNnode(int n){
    node* index = getNnode(n+1);
    index->pre->next = index->next;
    index->next->pre = index->pre;
    size--;
    delete index;

    return this;
  }

  linkedList* delnode(){
    delNnode(0); //del fist node
    //delNnode(size-1);//del last node

    return this;
  }

  ~linkedList(){
    while(size!=0){
      delnode();
    }

    delete start;
    delete end;
  }



};

int main(int argc, char const *argv[]) {
  linkedList a;
  for(int i =0;i<7;i++){
    a.insert(i);
  }
  a.printAll();
  a.insert(10,5);
  a.printAll();


  return 0;
}
